### PR TITLE
Add maintainer workflow, simplify git workflow

### PR DESCRIPTION
**Pull request by aradi**

### CODE REVIEW

File: docs/gitworkflow.rst [hunk 1]
```diff
@@ -20,11 +20,12 @@ developers are:
   code is clean, it is well documented, thoroughly tested, etc.), the feature
   branch is merged into the `master` branch.
 
-The main official public (upstream) repository only contains two branches: the
-branches `master` and `release`, latter containing tagged official releases.
-Some short living intermediate branches (e.g. `stage` and `hotfix`) may appear
-from time to time, but these are special purpose branches created by the
-administrators/release managers and are not being used for feature development.
+* In order to ease integration, feature branches should be **short living** and
+  pull requests should only contain a **reasonably small amount** of
```
Comment (awvwgk) [inline docs/gitworkflow.rst:24]: :+1: definitely worth highlighting this

File: docs/maintainerflow.rst [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+Following strategy should be applied when merging the pull request:
+
+- If the PR contains a **single commit** only and is based on current master, it
+  can be merged by fast-forward.
+
+- In all other cases, a merge commit should be created.
+
+- The code in the PR should be brought up to date with master by the contributor
+  before the review. The contributor is of course free to decide, whether she/he
```
Comment (awvwgk) [inline docs/maintainerflow.rst]: maybe use singular they here?

File: docs/maintainerflow.rst [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+The groups as listed above have following rights:
+
+- Repository administrators: They have all possible rights on the repositories,
+  including the possibility to grant rights to other developers.
+
+- Repository maintainers: They have write access to project repositories. They
+  can review and merge pull requests (PR). They can write into the corresponding
+  repository of an of contributors during the review process. They have access
+  to the test server, so that they can execute extended tests on the code in a
+  pull request.
```
Comment (awvwgk) [inline docs/maintainerflow.rst:40]: Would go well together with enabling branch protection for the default/master branch.

**Review by awvwgk — APPROVED** [review]
Looks good to me.

**Review by thomas-niehaus — APPROVED** [review]
Sounds reasonable.

File: docs/gitworkflow.rst [hunk 4]
```diff
@@ -20,11 +20,12 @@ developers are:
...
+  changes. Try to chop your implementation in small self containing changes and
```
Comment (bhourahine) [inline docs/gitworkflow.rst]: "into small"

File: docs/maintainerflow.rst [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+In order to ease the process of maintaining, the following strategy is applied
+for the development of DFTB+ and related projects.
+
+Roles
+=====
+
+We have three groups of developers with different access rights to the
+repositories:
+
+- Coordinating developers, repository administrators: They are responsible for
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: Coordinating developers (repository administrators):

File: docs/maintainerflow.rst [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  strategic decisions. Some of them also represent the project in certain
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: instead of 'in certain commercial collaborations' how about just 'externally' ?

File: docs/maintainerflow.rst [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  commercial collaborations.
+
+- Core developers, repository maintainers: Experienced, skilled developers
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: Core developers (repository maintainers):

File: docs/maintainerflow.rst [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  contributing regularly to the projects and having a good overview over the
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: overview of the

File: docs/maintainerflow.rst [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  project. They are often responsible for certain set of features within the
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: 'set' should be pluralised (sets) or 'a set of'

File: docs/maintainerflow.rst [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+=============
+
+The groups as listed above have following rights:
+
+- Repository administrators: They have all possible rights on the repositories,
+  including the possibility to grant rights to other developers.
+
+- Repository maintainers: They have write access to project repositories. They
+  can review and merge pull requests (PR). They can write into the corresponding
+  repository of an of contributors during the review process. They have access
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: "of contributors during the review of pull requests."

File: docs/maintainerflow.rst [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  to the test server, so that they can execute extended tests on the code in a
```
Comment (bhourahine) [inline docs/maintainerflow.rst:39]: extended validation of the code

File: docs/maintainerflow.rst [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  pull request.
+
+- Contributors: They have the usual read and fork rights as granted by GitHub.
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: ... granted by GitHub and  can of course submit pull requests.

File: docs/maintainerflow.rst [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+Merging pull requests
+=====================
+
+The following rules are currently applied when merging pull requests:
+
+- All changes to the master branch should be made via pull requests.
+
+- Pull requests must be approved by a maintainer before they can be merged. Pull
+  requests issued by a maintainer must be approved by an other maintainer.
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: another

File: docs/maintainerflow.rst [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+- Trivial pull requests can be merged by any maintainer, provided it had been
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: has been
Also which rule (there are 2 above)? 'provided self-submitted code has been approved by another maintainer'

File: docs/maintainerflow.rst [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+  approved according to the rule above.
+
+- Pull requests with non-trivial or substantial changes (major re-structuring,
+  changes in code behavior, large functionality additions, etc.) must be
+  discussed with and approved by an admin before merging.
+
+- Pull requests should pass the extended manually triggered buildbot tests (not
+  just the automatic travis tests) before they can be merged.
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: "automatic travis tests on GitHub"

File: docs/maintainerflow.rst [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+Following strategy should be applied when merging the pull request:
+
+- If the PR contains a **single commit** only and is based on current master, it
+  can be merged by fast-forward.
+
+- In all other cases, a merge commit should be created.
+
+- The code in the PR should be brought up to date with master by the contributor
+  before the review. The contributor is of course free to decide, whether she/he
+  does it via a merge (merging master into the feature branch) or via a rebase
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: does this via

File: docs/maintainerflow.rst [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+  (rebasing the feature branch on current master -- only for skilled
+  contributors!). However, once the review has started, no rebasing of the
```
Comment (bhourahine) [inline docs/maintainerflow.rst]: no further rebases of the feature branch should happen.

File: docs/maintainerflow.rst (hunk 3 above)
Comment (aradi) [reply docs/maintainerflow.rst:40]: :heavy_check_mark:

**Pull request merged**